src/Network/Transport/TCP/Internal.hs

```diff
@@ -209,3 +222,47 @@ recvExact sock len = go [] len
       if BS.null bs
         then throwIO (userError "recvExact: Socket closed")
         else go (bs : acc) (l - fromIntegral (BS.length bs))
+
+-- | Produce a HostName and ServiceName from a SockAddr. Only gives 'Just' for
+-- IPv4 addresses.
+decodeSockAddr :: N.SockAddr -> IO (Maybe (N.HostName, N.ServiceName))
+decodeSockAddr sockAddr = case sockAddr of
+  N.SockAddrInet port host -> do
+    hostString <- N.inet_ntoa host
```

Comment:
I'd like to allow the use of domain names. I suppose we could use `getnameinfo` to try and find the name... or, have `createTransport` resolve the name to an address when it starts up, and use that as the endpoint address.